thunder/transforms/cudagraph.py

```diff
@@ -35,63 +34,75 @@ def extract_descriptor(arg):
             return type(arg), None, None, arg
 
     dtypes, sizes, strides, non_tensor_args = zip(*map(extract_descriptor, args))
-    return ArgsDescriptor(dtypes, sizes, strides, non_tensor_args, args)
+    return ArgsDescriptor(dtypes, sizes, strides, non_tensor_args)
 
 
-@lru_cache
-def build_cuda_graph(
-    fn: Callable, args_descriptor: ArgsDescriptor, static_args_mask: tuple[bool, ...]
-) -> tuple[torch.cuda.CUDAGraph, Sequence[torch.Tensor | Any], Sequence[torch.Tensor | Any]]:
+class CUDAGraphRunner:
+    def __init__(self):
+        self.cuda_graph_cache = {}
+        self.python_callables = {}  # fn_name -> (callable. num_static_inputs)
+        self.trace_symbols = {}  # fn_name -> (bsyms, inputs, outputs)
+        self.name_counter = 1
 
-    def get_static_buffer(x):
+    def get_static_buffer(self, x):
         if isinstance(x, torch.Tensor):
             return torch.empty_like(x).copy_(x)
         return x
 
-    args = args_descriptor.args
-
-    # Warmup
-    torch.cuda.synchronize()
-    stream = torch.cuda.Stream()
-    stream.wait_stream(torch.cuda.current_stream())
-
-    with torch.cuda.stream(stream):
-        static_inputs = tuple(
-            get_static_buffer(arg) if not is_static else arg for arg, is_static in zip(args, static_args_mask)
-        )
-        for _ in range(3):
-            fn(*static_inputs)
-
-    stream.synchronize()
-    torch.cuda.current_stream().wait_stream(stream)
-    torch.cuda.synchronize()
-
-    # Record
-    # NOTE: We are using default private pool here, but it is possibly better to
-    # use a custom pool for better memory management. See CUDA Graphs Tree in
-    # PyTorch's Inductor: torch/_inductor/cudagraph_trees.py
-    # Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/view
-    graph = torch.cuda.CUDAGraph()
-    with torch.cuda.graph(graph, stream=stream):
-        static_outputs = fn(*static_inputs)
-
-    return graph, static_inputs, static_outputs
+    def build_cuda_graph(
+        self, fn: Callable, args: list[any], static_args_mask: tuple[bool, ...]
+    ) -> tuple[torch.cuda.CUDAGraph, Sequence[torch.Tensor | Any], Sequence[torch.Tensor | Any]]:
```

Comment:
What is the benefit of having this as a method and not as a free function?

Reply:
My tinking was that having the ability to override the buffer allocation (get_static_buffer) could get useful. We could have it as a callback instead but then it might just as well be a method.